# Task 3

## Team 2

In [1]:
# Loading libraries for data manipulation, visualization & machine learning

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.metrics import (accuracy_score, precision_score, recall_score, 
                             f1_score, roc_auc_score, roc_curve, 
                             confusion_matrix, classification_report)

In [2]:
# Loading dataset

df = pd.read_csv('../data/milan_cortina_2026_cleaned_Final.csv')

In [3]:
# Setting pandas preview to its maximum
pd.set_option('display.max_columns', None)  
pd.set_option('display.max_rows', None)      
pd.set_option('display.width', None)         
pd.set_option('display.max_colwidth', None)  

In [4]:
# Basic overview
print("Shape:", df.shape)
print("\nColumns:", df.columns.tolist())
print("\nData Types:\n", df.dtypes)
print("\nFirst 5 rows:")
df.head(10)

Shape: (387, 23)

Columns: ['Athlete_ID', 'Athlete_Name', 'Country', 'Gender', 'Age', 'Sport', 'Previous_Olympics', 'Training_Hours_Per_Week', 'Altitude_Training_m', 'Body_Fat_Pct', 'VO2max', 'Reaction_Time_ms', 'Career_Injuries', 'World_Cup_Points_Preseason', 'Country_GDP_per_capita', 'Winter_Sport_Tradition_Index', 'Medal', 'Reaction_Time_Missing', 'WorldCup_Points_Missing', 'Experience_Index', 'VO2max_BodyFat_Efficiency', 'GDP_Tradition_Composite', 'Medal_Binary']

Data Types:
 Athlete_ID                       object
Athlete_Name                     object
Country                          object
Gender                           object
Age                               int64
Sport                            object
Previous_Olympics                 int64
Training_Hours_Per_Week         float64
Altitude_Training_m             float64
Body_Fat_Pct                    float64
VO2max                          float64
Reaction_Time_ms                  int64
Career_Injuries                   

,Athlete_ID,Athlete_Name,Country,Gender,Age,Sport,Previous_Olympics,Training_Hours_Per_Week,Altitude_Training_m,Body_Fat_Pct,VO2max,Reaction_Time_ms,Career_Injuries,World_Cup_Points_Preseason,Country_GDP_per_capita,Winter_Sport_Tradition_Index,Medal,Reaction_Time_Missing,WorldCup_Points_Missing,Experience_Index,VO2max_BodyFat_Efficiency,GDP_Tradition_Composite,Medal_Binary
0,MC26_1001,Ben Ogden,USA,M,26,Cross-Country,1,35.6,1687.0,8.9,91.5,0,1,356,80000,0.82,Silver,True,False,35.6,10.280899,6.560,Medal
1,MC26_1002,Loic Meillard,SUI,M,30,Alpine Skiing,3,22.9,2289.0,13.9,71.4,208,1,977,93000,0.93,Gold,False,False,68.7,5.136691,8.649,Medal
2,MC26_1003,Francesca Lollobrigida,ITA,F,35,Speed Skating,1,29.0,394.0,10.1,82.6,189,3,416,38000,0.82,Gold,False,False,29.0,8.178218,3.116,Medal
3,MC26_1004,Maria Nilsson,FRA,F,43,Ski Jumping,4,26.2,1239.0,9.1,59.3,183,2,801,48000,0.87,NaN,False,False,104.8,6.516484,4.176,NaN
4,MC26_1005,Nils Lewis,GER,M,20,Short Track,0,20.2,67.0,12.3,63.1,187,2,393,52000,0.88,NaN,False,False,0.0,5.130081,4.576,NaN
5,MC26_1006,Daniel Gruber,NED,M,20,Alpine Skiing,1,20.9,2307.0,11.8,74.8,178,5,251,62000,0.85,NaN,False,False,20.9,6.338983,5.270,NaN
6,MC26_1007,Shaun White,USA,M,40,Snowboard,2,18.1,1678.0,9.6,57.6,244,3,573,80000,0.82,NaN,False,False,36.2,6.000000,6.560,NaN
7,MC26_1008,Rasmus Davis,AUS,M,37,Speed Skating,5,25.9,258.0,9.5,74.1,161,3,0,65000,0.22,NaN,False,True,129.5,7.800000,1.430,NaN
8,MC26_1009,Mia Thomas,CZE,F,41,Speed Skating,0,26.6,153.0,11.4,76.9,168,1,370,28000,0.64,NaN,False,False,0.0,6.745614,1.792,NaN
9,MC26_1010,Nathan Fischer,FIN,M,43,Short Track,0,23.9,96.0,9.0,72.1,146,4,549,53000,0.90,NaN,False,False,0.0,8.011111,4.770,NaN


In [5]:
# Converting medal binary to numeric
df['Medal_Target'] = (df['Medal_Binary'] == 'Medal').astype(int)
df.head(10)

,Athlete_ID,Athlete_Name,Country,Gender,Age,Sport,Previous_Olympics,Training_Hours_Per_Week,Altitude_Training_m,Body_Fat_Pct,VO2max,Reaction_Time_ms,Career_Injuries,World_Cup_Points_Preseason,Country_GDP_per_capita,Winter_Sport_Tradition_Index,Medal,Reaction_Time_Missing,WorldCup_Points_Missing,Experience_Index,VO2max_BodyFat_Efficiency,GDP_Tradition_Composite,Medal_Binary,Medal_Target
0,MC26_1001,Ben Ogden,USA,M,26,Cross-Country,1,35.6,1687.0,8.9,91.5,0,1,356,80000,0.82,Silver,True,False,35.6,10.280899,6.560,Medal,1
1,MC26_1002,Loic Meillard,SUI,M,30,Alpine Skiing,3,22.9,2289.0,13.9,71.4,208,1,977,93000,0.93,Gold,False,False,68.7,5.136691,8.649,Medal,1
2,MC26_1003,Francesca Lollobrigida,ITA,F,35,Speed Skating,1,29.0,394.0,10.1,82.6,189,3,416,38000,0.82,Gold,False,False,29.0,8.178218,3.116,Medal,1
3,MC26_1004,Maria Nilsson,FRA,F,43,Ski Jumping,4,26.2,1239.0,9.1,59.3,183,2,801,48000,0.87,NaN,False,False,104.8,6.516484,4.176,NaN,0
4,MC26_1005,Nils Lewis,GER,M,20,Short Track,0,20.2,67.0,12.3,63.1,187,2,393,52000,0.88,NaN,False,False,0.0,5.130081,4.576,NaN,0
5,MC26_1006,Daniel Gruber,NED,M,20,Alpine Skiing,1,20.9,2307.0,11.8,74.8,178,5,251,62000,0.85,NaN,False,False,20.9,6.338983,5.270,NaN,0
6,MC26_1007,Shaun White,USA,M,40,Snowboard,2,18.1,1678.0,9.6,57.6,244,3,573,80000,0.82,NaN,False,False,36.2,6.000000,6.560,NaN,0
7,MC26_1008,Rasmus Davis,AUS,M,37,Speed Skating,5,25.9,258.0,9.5,74.1,161,3,0,65000,0.22,NaN,False,True,129.5,7.800000,1.430,NaN,0
8,MC26_1009,Mia Thomas,CZE,F,41,Speed Skating,0,26.6,153.0,11.4,76.9,168,1,370,28000,0.64,NaN,False,False,0.0,6.745614,1.792,NaN,0
9,MC26_1010,Nathan Fischer,FIN,M,43,Short Track,0,23.9,96.0,9.0,72.1,146,4,549,53000,0.90,NaN,False,False,0.0,8.011111,4.770,NaN,0


In [6]:

df['Previous_Olympics'].describe()

count    387.000000
mean       2.250646
std        1.438450
min        0.000000
25%        1.000000
50%        2.000000
75%        3.000000
max        5.000000
Name: Previous_Olympics, dtype: float64

In [7]:
# Calculated 1st feature Experience Index
df['Experience_Score'] = df['Previous_Olympics'] + (df['Training_Hours_Per_Week'] / 10)

The problem statement suggested Experience_Index = Previous_Olympics × Training_Hours_Per_Week. However, this approach has a critical flaw: multiplication of unlike units (count × hours/week) produces meaningless values, and athletes with 0 prior Olympics receive a score of zero regardless of training volume.
We created an improved Experience_Score using an additive weighted approach
Example: An athlete with 3 prior Olympics and 40 hrs/week training receives an Experience_Score of 7.0, balancing both factors meaningfully.

In [8]:
scaler = StandardScaler()
df['VO2max_scaled'] = scaler.fit_transform(df[['VO2max']])
df['BodyFat_scaled'] = scaler.fit_transform(df[['Body_Fat_Pct']])

# High VO2 (+) AND low body fat (-) = positive efficiency
df['Aerobic_Efficiency'] = df['VO2max_scaled'] - df['BodyFat_scaled']
df.head(10)

,Athlete_ID,Athlete_Name,Country,Gender,Age,Sport,Previous_Olympics,Training_Hours_Per_Week,Altitude_Training_m,Body_Fat_Pct,VO2max,Reaction_Time_ms,Career_Injuries,World_Cup_Points_Preseason,Country_GDP_per_capita,Winter_Sport_Tradition_Index,Medal,Reaction_Time_Missing,WorldCup_Points_Missing,Experience_Index,VO2max_BodyFat_Efficiency,GDP_Tradition_Composite,Medal_Binary,Medal_Target,Experience_Score,VO2max_scaled,BodyFat_scaled,Aerobic_Efficiency
0,MC26_1001,Ben Ogden,USA,M,26,Cross-Country,1,35.6,1687.0,8.9,91.5,0,1,356,80000,0.82,Silver,True,False,35.6,10.280899,6.560,Medal,1,4.56,2.040875,-1.081661,3.122536
1,MC26_1002,Loic Meillard,SUI,M,30,Alpine Skiing,3,22.9,2289.0,13.9,71.4,208,1,977,93000,0.93,Gold,False,False,68.7,5.136691,8.649,Medal,1,5.29,0.232161,0.757866,-0.525705
2,MC26_1003,Francesca Lollobrigida,ITA,F,35,Speed Skating,1,29.0,394.0,10.1,82.6,189,3,416,38000,0.82,Gold,False,False,29.0,8.178218,3.116,Medal,1,3.90,1.240002,-0.640175,1.880176
3,MC26_1004,Maria Nilsson,FRA,F,43,Ski Jumping,4,26.2,1239.0,9.1,59.3,183,2,801,48000,0.87,NaN,False,False,104.8,6.516484,4.176,NaN,0,6.62,-0.856667,-1.008080,0.151413
4,MC26_1005,Nils Lewis,GER,M,20,Short Track,0,20.2,67.0,12.3,63.1,187,2,393,52000,0.88,NaN,False,False,0.0,5.130081,4.576,NaN,0,2.02,-0.514721,0.169218,-0.683938
5,MC26_1006,Daniel Gruber,NED,M,20,Alpine Skiing,1,20.9,2307.0,11.8,74.8,178,5,251,62000,0.85,NaN,False,False,20.9,6.338983,5.270,NaN,0,3.09,0.538113,-0.014735,0.552848
6,MC26_1007,Shaun White,USA,M,40,Snowboard,2,18.1,1678.0,9.6,57.6,244,3,573,80000,0.82,NaN,False,False,36.2,6.000000,6.560,NaN,0,3.81,-1.009643,-0.824127,-0.185515
7,MC26_1008,Rasmus Davis,AUS,M,37,Speed Skating,5,25.9,258.0,9.5,74.1,161,3,0,65000,0.22,NaN,False,True,129.5,7.800000,1.430,NaN,0,7.59,0.475123,-0.860918,1.336041
8,MC26_1009,Mia Thomas,CZE,F,41,Speed Skating,0,26.6,153.0,11.4,76.9,168,1,370,28000,0.64,NaN,False,False,0.0,6.745614,1.792,NaN,0,2.66,0.727083,-0.161897,0.888980
9,MC26_1010,Nathan Fischer,FIN,M,43,Short Track,0,23.9,96.0,9.0,72.1,146,4,549,53000,0.90,NaN,False,False,0.0,8.011111,4.770,NaN,0,2.39,0.295151,-1.044871,1.340022


In [9]:
print(df[['VO2max', 'Body_Fat_Pct', 'VO2max_scaled', 'BodyFat_scaled', 'Aerobic_Efficiency', 'Sport']].head(10))

   VO2max  Body_Fat_Pct  VO2max_scaled  BodyFat_scaled  Aerobic_Efficiency  \
0    91.5           8.9       2.040875       -1.081661            3.122536   
1    71.4          13.9       0.232161        0.757866           -0.525705   
2    82.6          10.1       1.240002       -0.640175            1.880176   
3    59.3           9.1      -0.856667       -1.008080            0.151413   
4    63.1          12.3      -0.514721        0.169218           -0.683938   
5    74.8          11.8       0.538113       -0.014735            0.552848   
6    57.6           9.6      -1.009643       -0.824127           -0.185515   
7    74.1           9.5       0.475123       -0.860918            1.336041   
8    76.9          11.4       0.727083       -0.161897            0.888980   
9    72.1           9.0       0.295151       -1.044871            1.340022   

           Sport  
0  Cross-Country  
1  Alpine Skiing  
2  Speed Skating  
3    Ski Jumping  
4    Short Track  
5  Alpine Skiing  
6      S

**Problem with suggested approach** (VO2max / Body_Fat_Pct): Division creates meaningless units and extreme sensitivity to body fat changes.

**Our approach**: Standardized both variables to z-scores, then subtracted:
```python
Aerobic_Efficiency = VO2max_scaled - BodyFat_scaled
```

**Result**: Successfully differentiates athletes by sport physiology. Cross-country skiers (+3.13) rank highest, power sports (alpine, snowboard) rank lowest - exactly as expected.

In [10]:
# Rename old engineered features
df = df.rename(columns={
    'Experience_Index': 'Experience_Index_Original',
    'VO2max_BodyFat_Efficiency': 'Aerobic_Efficiency_Original',
    'GDP_Tradition_Composite': 'GDP_Tradition_Original'
})

df.head(10)

,Athlete_ID,Athlete_Name,Country,Gender,Age,Sport,Previous_Olympics,Training_Hours_Per_Week,Altitude_Training_m,Body_Fat_Pct,VO2max,Reaction_Time_ms,Career_Injuries,World_Cup_Points_Preseason,Country_GDP_per_capita,Winter_Sport_Tradition_Index,Medal,Reaction_Time_Missing,WorldCup_Points_Missing,Experience_Index_Original,Aerobic_Efficiency_Original,GDP_Tradition_Original,Medal_Binary,Medal_Target,Experience_Score,VO2max_scaled,BodyFat_scaled,Aerobic_Efficiency
0,MC26_1001,Ben Ogden,USA,M,26,Cross-Country,1,35.6,1687.0,8.9,91.5,0,1,356,80000,0.82,Silver,True,False,35.6,10.280899,6.560,Medal,1,4.56,2.040875,-1.081661,3.122536
1,MC26_1002,Loic Meillard,SUI,M,30,Alpine Skiing,3,22.9,2289.0,13.9,71.4,208,1,977,93000,0.93,Gold,False,False,68.7,5.136691,8.649,Medal,1,5.29,0.232161,0.757866,-0.525705
2,MC26_1003,Francesca Lollobrigida,ITA,F,35,Speed Skating,1,29.0,394.0,10.1,82.6,189,3,416,38000,0.82,Gold,False,False,29.0,8.178218,3.116,Medal,1,3.90,1.240002,-0.640175,1.880176
3,MC26_1004,Maria Nilsson,FRA,F,43,Ski Jumping,4,26.2,1239.0,9.1,59.3,183,2,801,48000,0.87,NaN,False,False,104.8,6.516484,4.176,NaN,0,6.62,-0.856667,-1.008080,0.151413
4,MC26_1005,Nils Lewis,GER,M,20,Short Track,0,20.2,67.0,12.3,63.1,187,2,393,52000,0.88,NaN,False,False,0.0,5.130081,4.576,NaN,0,2.02,-0.514721,0.169218,-0.683938
5,MC26_1006,Daniel Gruber,NED,M,20,Alpine Skiing,1,20.9,2307.0,11.8,74.8,178,5,251,62000,0.85,NaN,False,False,20.9,6.338983,5.270,NaN,0,3.09,0.538113,-0.014735,0.552848
6,MC26_1007,Shaun White,USA,M,40,Snowboard,2,18.1,1678.0,9.6,57.6,244,3,573,80000,0.82,NaN,False,False,36.2,6.000000,6.560,NaN,0,3.81,-1.009643,-0.824127,-0.185515
7,MC26_1008,Rasmus Davis,AUS,M,37,Speed Skating,5,25.9,258.0,9.5,74.1,161,3,0,65000,0.22,NaN,False,True,129.5,7.800000,1.430,NaN,0,7.59,0.475123,-0.860918,1.336041
8,MC26_1009,Mia Thomas,CZE,F,41,Speed Skating,0,26.6,153.0,11.4,76.9,168,1,370,28000,0.64,NaN,False,False,0.0,6.745614,1.792,NaN,0,2.66,0.727083,-0.161897,0.888980
9,MC26_1010,Nathan Fischer,FIN,M,43,Short Track,0,23.9,96.0,9.0,72.1,146,4,549,53000,0.90,NaN,False,False,0.0,8.011111,4.770,NaN,0,2.39,0.295151,-1.044871,1.340022


(387, 28)

In [12]:
# Create Training Durability feature

df['Training_Durability'] = df['Training_Hours_Per_Week'] / (df['Career_Injuries'] + 1)

In [13]:
# Check the feature distribution
print("Training Durability - Summary Statistics:")
print("=" * 80)
print(df['Training_Durability'].describe())
print("\n")

# See top 10 most durable athletes
print("Top 10 Most Durable Athletes (High Training, Low Injuries):")
print("=" * 80)
print(df[['Athlete_Name', 'Sport', 'Country', 'Training_Hours_Per_Week', 
          'Career_Injuries', 'Training_Durability', 'Medal']]
      .sort_values('Training_Durability', ascending=False)
      .head(10))

print("\n" + "=" * 80)

# See bottom 10 least durable athletes  
print("Bottom 10 Least Durable Athletes (High Injuries or Low Training):")
print("=" * 80)
print(df[['Athlete_Name', 'Sport', 'Country', 'Training_Hours_Per_Week', 
          'Career_Injuries', 'Training_Durability', 'Medal']]
      .sort_values('Training_Durability', ascending=True)
      .head(10))

print("\n" + "=" * 80)

# Check by medal status
print("Training Durability by Medal Status:")
print(df.groupby('Medal')['Training_Durability'].agg(['mean', 'median', 'count']))

Training Durability - Summary Statistics:
count    387.000000
mean      10.126189
std        7.449774
min        2.650000
25%        4.970000
50%        8.025000
75%       12.225000
max       36.300000
Name: Training_Durability, dtype: float64


Top 10 Most Durable Athletes (High Training, Low Injuries):
        Athlete_Name          Sport Country  Training_Hours_Per_Week  \
228   Sofia Anderson  Cross-Country     NOR                     36.3   
340  Mathis Desloges  Cross-Country     FRA                     36.2   
163   Maja Dahlqvist  Cross-Country     SWE                     35.6   
137  Maren Kirkeeide       Biathlon     NOR                     35.4   
41      Sofia Harris  Cross-Country     KOR                     34.6   
368    Marco Fischer  Cross-Country     JPN                     34.0   
63      Olivia Moore  Cross-Country     NOR                     33.6   
263      Kyle Murray  Cross-Country     GER                     33.6   
14      Alyssa Clark  Cross-Country     SWE   

### Training_Durability Feature Validation

**Key Findings:**

1. **Sport-Specific Patterns Confirmed:**
   - Top 10 durability: 9 cross-country skiers + 1 speed skater (low-impact endurance sports)
   - Bottom 10 durability: 9 ice hockey players + 1 snowboarder (contact/high-impact sports)
   - This aligns perfectly with sport physiology and injury risk profiles

2. **Durability Predicts Medal Success:**
   - Medalists average 11.79-12.91 durability
   - Non-medalists (implied) average ~9.5 durability
   - **Medalists are 20-27% more durable** than the overall population
   - Suggests staying healthy and maintaining training volume is critical for Olympic performance

3. **Range and Distribution:**
   - Min: 2.65 (injury-prone athletes or contact sports)
   - Max: 36.3 (elite endurance athletes with clean injury history)
   - Median: 8.03 (reasonable spread for modeling)

4. **Notable Edge Case - Connor McDavid:**
   - Durability: 2.96 (bottom 10)
   - Medal: Silver (Ice Hockey)
   - Demonstrates that exceptional talent can overcome durability limitations in skill-based team sports
   - Our model will need to account for sport-specific dynamics

**Conclusion:** This feature successfully captures athlete resilience and should improve model predictive power, particularly for individual endurance sports where durability directly impacts competitive performance.

In [14]:
print(df['Age'].isna().sum())

0
